### PR TITLE
add msg to _assert_async

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134823
* #134819
* #134818
* __->__ #134813

cc @malfet

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Review by eqy — APPROVED** [review]

File: aten/src/ATen/native/cuda/TensorCompare.cu [hunk 1]
```diff
@@ -101,33 +101,42 @@ REGISTER_DISPATCH(clamp_scalar_stub, &clamp_scalar_kernel_impl);
...
-    _assert_async_cuda_kernel<<<1, 1, 0, stream>>>(self.const_data_ptr<scalar_t>());
+    _assert_async_cuda_kernel<<<1, 1, 0, stream>>>(self.const_data_ptr<scalar_t>(), msg);
     C10_CUDA_KERNEL_LAUNCH_CHECK();
   });
 }
 
 // TODO (tmanlaibaatar) Ignore assert msg for now
-void _assert_async_msg_cuda(const Tensor& self_tensor, c10::string_view assert_msg) {
-  _assert_async_cuda(self_tensor);
+void _assert_async_cuda(const Tensor& self_tensor) {
```
Comment (albanD) [inline aten/src/ATen/native/cuda/TensorCompare.cu:138]: The TODO above should be removed

File: aten/src/ATen/native/cuda/TensorCompare.cu [hunk 2]
```diff
@@ -101,33 +101,42 @@ REGISTER_DISPATCH(clamp_scalar_stub, &clamp_scalar_kernel_impl);
...
+};
 template <typename scalar_t>
-__global__ void _assert_async_cuda_kernel(const scalar_t* input) {
-  CUDA_KERNEL_ASSERT(input[0] != 0);
+__global__ void _assert_async_cuda_kernel(const scalar_t* input, Msg msg) {
+  CUDA_KERNEL_ASSERT_MSG(input[0] != 0, msg.msg);
 }
 
-__global__ void _assert_async_cuda_kernel(const c10::complex<float>* input) {
+__global__ void _assert_async_cuda_kernel(const c10::complex<float>* input, Msg msg) {
```
Comment (albanD) [inline aten/src/ATen/native/cuda/TensorCompare.cu:113]: Don't you need to update the macro to take in the "Msg" here and below?

File: c10/macros/Macros.h [hunk 3]
```diff
@@ -372,6 +373,14 @@ __host__ __device__
                static_cast<unsigned>(__LINE__)), \
            0);                                   \
   }
+#define CUDA_KERNEL_ASSERT_MSG(cond, msg)        \
+  if (C10_UNLIKELY(!(cond))) {                   \
+    (void)(_wassert(                             \
+               _CRT_WIDE(msg),                   \
```
Comment (albanD) [inline c10/macros/Macros.h:379]: I think you can only apply this to a string, not variable. Hence the windows build failure.

**Review by albanD — COMMENTED** [review]
It was very surprising that someone went through the trouble to add the overload but not the impl haha

Windows failures are real though

File: c10/macros/Macros.h (hunk 3 above)
Comment (Chillee) [reply c10/macros/Macros.h:379]: >:(

File: c10/macros/Macros.h [hunk 4]
```diff
@@ -413,13 +420,19 @@ __host__ __device__
...
 #else
 #define CUDA_KERNEL_ASSERT(cond)                                         \
   if (C10_UNLIKELY(!(cond))) {                                           \
     __assert_fail(                                                       \
         #cond, __FILE__, static_cast<unsigned int>(__LINE__), __func__); \
   }
+#define CUDA_KERNEL_ASSERT_MSG(cond, msg)                              \
+  if (C10_UNLIKELY(!(cond))) {                                         \
+    __assert_fail(                                                     \
+        msg, __FILE__, static_cast<unsigned int>(__LINE__), __func__); \
```
Comment (albanD) [inline c10/macros/Macros.h:434]:
```suggestion
        msg, _CRT_WIDE(__FILE__,) static_cast<unsigned int>(__LINE__), __func__); \
```

I guess for consistency? But not sure how much cuda testing we have on windows running in CI :p

**Review by albanD — APPROVED** [review]
SGTM!

File: c10/macros/Macros.h (hunk 3 above)
Comment (Chillee) [reply c10/macros/Macros.h]: I'm just going to not throw the device-side assert on windows. A brief search did not show up any easy ways of converting a char* to a wchar*.

**Comment by yinghai** [comment]
who is calling `_assert_async_msg_cuda` out of curiosity.

**Comment by Chillee** [comment]
@yinghai me! https://github.com/pytorch/pytorch/pull/134818/files#diff-67981d6b129d1343a92ce9fac7b0e9361cbad3b45b298b1237fb6e3e3ebf4499R590

But in theory this could be generically useful.

**Pull request closed without merging**